src/gps/nmea_parser.cc

```diff
@@ -0,0 +1,137 @@
+{
+    if (!parse(line, "$GNZDA", 6)) {
+        return false;
+    }
+
+    try {
+        parseTimeField(m_Fields[0], time);
+        time.tm_mday = stoi(m_Fields[1]);
+        time.tm_mon = stoi(m_Fields[2]) - 1;
+        time.tm_year = stoi(m_Fields[3]) - 1900;
+
+        // This field is available in GNU/BSD but is not strict ISO C
+        int hoursOff = m_Fields[4].empty() ? 0 : stoi(m_Fields[4]);
+        int minutesOff = m_Fields[5].size() <= 3 ? 0 : stoi(m_Fields[5].substr(0, 2));
+        time.tm_gmtoff = 3600 * hoursOff + 60 * minutesOff;
+    } catch (std::invalid_argument &e) {
+        throw NMEAError(e.what());
+    } catch (std::out_of_range &e) {
+        throw NMEAError(e.what());
+    }
+
+    return true;
+}
+
+bool
+NMEAParser::parseCoordinates(const std::string &line, GPSData &data)
+{
+    using meter_t = units::length::meter_t;
+    using degree_t = units::angle::degree_t;
+    using arcminute_t = units::angle::arcminute_t;
+
+    if (!parse(line, "$GNGGA", 9)) {
+        return false;
+    }
+
+    try {
+        parseTimeField(m_Fields[0], data.time); // UTC time
+        data.milliseconds = stoi(m_Fields[0].substr(7, 2));
+
+        data.coordinate.lat = degree_t(std::stod(m_Fields[1].substr(0, 2))) +
+                              arcminute_t(std::stod(m_Fields[1].substr(2, 8)));
+        if (m_Fields[2][0] == 'W') {
+            data.coordinate.lat = -data.coordinate.lat;
+        }
+        data.coordinate.lon = degree_t(std::stod(m_Fields[3].substr(0, 3))) +
+                              arcminute_t(std::stod(m_Fields[3].substr(3, 9)));
+        if (m_Fields[4][0] == 'S') {
+            data.coordinate.lon = -data.coordinate.lon;
+        }
+        data.gpsQuality = static_cast<GPSQuality>(std::stoi(m_Fields[5]));
+        data.numberOfSatellites = std::stoi(m_Fields[6]);
+        data.horizontalDilution = meter_t{ std::stod(m_Fields[7]) };
+        data.coordinate.height = meter_t{ std::stod(m_Fields[8]) };
+    } catch (std::invalid_argument &e) {
+        throw NMEAError(e.what());
+    } catch (std::out_of_range &e) {
+        throw NMEAError(e.what());
+    }
+
+    return true;
+}
+
+bool
+NMEAParser::parse(const std::string &line, const std::string &sentenceId,
+                  size_t numFields)
+{
+    BOB_ASSERT(numFields <= m_Fields.size());
+
+    std::string field;
+    std::istringstream lineStream{ line };
+
+    // The first field indicates what kind of message we have
+    if (!getline(lineStream, field, ',')) {
+        return false;
+    }
+    if (field.empty() || field[0] != '$') {
+        throw NMEAError{ "Sentence doesn't begin with $" };
+    }
+    if (field != sentenceId) {
+        return false;
+    }
+
+    if (line.size() < 3 || line[line.size() - 3] != '*') {
+        throw NMEAError{ "Missing checksum" };
+    }
+
+    // Check that the checksum matches
+    char *p;
+    long reportedChecksum = std::strtoul(&line[line.size() - 2], &p, 16);
```

Comment:
This is a pretty hardcore ``std::strtoul`` invocation (I definitely had to look at the manual) - maybe add a comment to explain that you're parsing in hex and that you use the p to detect failure (I figure you're doing it because 0 could also be a valid checksum but from the documentation this still seems suspect - I see no reference to this being a guaranteed way to signal an error). In short, I *think* you should use ``std::stoul`` instead (https://en.cppreference.com/w/cpp/string/basic_string/stoul) and just check for ``std::invalid_argument`` and ``std::out_of_range``